File: Fillər.py
# https://basecamp.eolymp.com/az/problems/9630
def f(x, y, a):
    if x >= 2:
        return a[y] - a[x - 2]
    else:
        return a[y]

def main():
    n, m = map(int, input().split())
    
    a = [0] * (2 * n + 2)
    ans = [0] * (2 * n + 2)
    ans2 = [0] * (2 * n + 2)
    cur = 0
    
    for _ in range(m):
        x, y = map(int, input().split())
        
        pos1 = x + y
        pos2 = x - y + n
        
        if ans[pos1] == 0:
            if pos1 > n + 1:
                cur += 2 * n - pos1 + 1
            else:
                cur += pos1 - 1
            ans[pos1] = a[pos1] = 1
        
        if ans2[pos2] == 0:
            if x - y < 0:
                cur += n + (x - y)
            else:
                cur += n - (x - y)
            ans2[pos2] = 1
    
    for i in range(2, 2 * n + 1):
        a[i] += a[i - 2]
    
    for i in range(n, 0, -1):
        if not ans2[n + i - 1]:
            continue
        cur -= f(i + 1, 2 * n - i + 1, a)
    
    for i in range(2, n + 1):
        if ans2[n - i + 1] == 0:
            continue
        cur -= f(i + 1, 2 * n - i + 1, a)
    
    print(n * n - cur)
    
if __name__ == "__main__":
    main()

File: Fillər.cpp
// https://basecamp.eolymp.com/az/problems/9630

#include <bits/stdc++.h>

using namespace std;
long long a[4000010], ans[4000010], ans2[4000010], n, m, cur;
long long f(long long x, long long y)
{
    return a[y] - a[x - 2] * (x >= 2);
}
int main()
{
    ios_base::sync_with_stdio(false), cout.tie(0), cin.tie(0);
    cin >> n >> m;
    long long x, y, pos1, pos2;
    for (int i = 1; i <= m; ++i)
    {
        cin >> x >> y;
        pos1 = (pos2 = x - y + n, x + y);
        if (ans[pos1] == 0)
        {
            if (pos1 > n + 1)
                cur += ((n << 1) - pos1 + 1);
            else
                cur += (pos1 - 1);
            ans[pos1] = a[pos1] = 1;
        }
        if (ans2[pos2] == 0)
        {
            if (x - y < 0)
                cur += (n + (x - y));
            else
                cur += (n - (x - y));
            ans2[pos2] = 1;
        }
    }
    for (int i = 2; i <= (n << 1); ++i)
        a[i] += a[i - 2];
    for (int i = n; i; --i)
        cur -= f(i + 1, (n << 1) - i + 1) * bool(ans2[n + i - 1]);
    for (int i = 2; i <= n; ++i)
        cur -= f(i + 1, (n << 1) - i + 1) * bool(ans2[n - i + 1]);
    cout << n * n - cur << '\n';
}